### PR TITLE
[FLINK-34487][ci] Adds Python Wheels nightly GHA workflow

### PR DESCRIPTION
**Pull request by morazow**
## What is the purpose of the change

Integrates Python Wheels build into the GitHub Actions (GHA) nightly workflow.

## Brief change log

- Adds Python Wheels GHA workflow, `template.python-wheels-ci.yml`
- Integrates the Python Wheels workflow to the nightly GHA workflow

## Verifying this change

This changes does not change code. 

But the successful nightly GHA run should verify the change.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: .github/workflows/template.python-wheels-ci.yml
Comment (morazow) [inline .github/workflows/template.python-wheels-ci.yml:41]: This is slightly different than the Azure pipelines build. There `Agent.JobName` is used.

```
- task: PublishPipelineArtifact@0
        inputs:
          artifactName: 'wheel_$(Agent.OS)_$(Agent.JobName)'
          targetPath: 'flink-python/dist'
```

Could you please suggest how could we make them similar?

**Review by affo — APPROVED** [review]
Thanks @morazow !

Looks like a 1:1 port of https://github.com/apache/flink/blob/master/tools/azure-pipelines/build-python-wheels.yml which is what the issue is there for.

LGTM 👍 

Waiting for other, more authoritative people, to reply to your questions above :)

File: .github/workflows/nightly.yml
Comment (XComp) [inline .github/workflows/nightly.yml:97]: I'm wondering whether we should just inline the two jobs. Adding them as a template makes the workflow appear on the left side of the workflow list in GitHub Actions (analogously to [the entry for the template.flink-ci.yml](https://github.com/apache/flink/actions/workflows/template.flink-ci.yml) without any additional utilization.

One argument to keep the template workflow could be that we would be able to enable `workflow_dispatch` for the wheel generation. But I'm not aware of such a use-case. :thinking: 

WDYT? :thinking:

File: .github/workflows/template.python-wheels-ci.yml
Comment (XComp) [reply .github/workflows/template.python-wheels-ci.yml:41]: What are you referring to with "them"? The `name`, I guess?

I'm wondering whether we should rely on what's provided in [apache/flink:.github/workflows/template.flink-ci.yml:106}(https://github.com/apache/flink/blob/master/.github/workflows/template.flink-ci.yml#L106) with the stringified workflow name. WDYT?

File: .github/workflows/template.python-wheels-ci.yml
Comment (XComp) [inline .github/workflows/template.python-wheels-ci.yml:37]: Would it make sense to also compress it here? Or is there a reason why compression is not enabled :thinking:

File: .github/workflows/template.python-wheels-ci.yml
Comment (XComp) [inline .github/workflows/template.python-wheels-ci.yml:51]: I'm wondering whether we should also add the explicit python installation to Linux. To have both jobs use the same Python version. Have you checked that this is working on the GHA runners (i.e. is Python installed on Linux but not MacOS)?

**Review by XComp — COMMENTED** [review]
Thanks for the PR, @morazow and @affo for looking into it. I passed over it as well and have a few comments. PTAL :-)

**Comment by XComp** [comment]
fyi: I won't be responsive till the end of April at least

File: .github/workflows/template.python-wheels-ci.yml
Comment (morazow) [reply .github/workflows/template.python-wheels-ci.yml:41]: Hey @XComp,

I meant to make the python-wheels-ci GitHub workflow and Azure pipelines.

Here the main concern is to keep the Python Wheels artifact names similar. Since the Azure pipelines used `Agent.JobName` it would be fine to use the stringrified workflow name, with `wheel` prefixed.

Would that be good option?

File: .github/workflows/template.python-wheels-ci.yml
Comment (morazow) [reply .github/workflows/template.python-wheels-ci.yml:51]: Yes good idea, I took the same steps from Azure pipelines. But I agree having same Python version on both Linux/MacOs jobs is good idea. I will change workflow.

I didn't check which version is installed in the Linux VM

**Comment by morazow** [comment]
Thanks @XComp, I am going to address your suggestions. Please have a look once you are back

**Comment by morazow** [comment]
@flinkbot run azure

File: .github/workflows/nightly.yml
Comment (XComp) [inline .github/workflows/nightly.yml:131]: Shouldn't we add the `linux` phrase here somewhere? Otherwise, we would have the same build artifact name twice (for macos and linux). Or am I missing something?

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks @morazow for addressing my comments. The PR looks good already. I have a few comments, though. PTAL

The commit message prefixes of the PR are a bit confusing. `[hotfix]` indicates that you want to have each of those commits as a separate commit. But we should squash them all together in the end. If it was actually your intention to squash everything before merging, you might want to use some other than `[hotfix]` (e.g. `[review]`) to mark the commits properly. :-)

**Comment by morazow** [comment]
Thanks @XComp, going through your suggestions, I will address them, thanks 👍 

No indeed, I am going to squash them, it should be single commit

File: .github/workflows/nightly.yml
Comment (morazow) [reply .github/workflows/nightly.yml:121]: Good catch, it is missing from env, added

File: .github/workflows/nightly.yml
Comment (XComp) [inline .github/workflows/nightly.yml:104]: That line is obsolete since you have this parameter already defined in the `.include` map below

File: .github/workflows/nightly.yml
Comment (XComp) [inline .github/workflows/nightly.yml:103]: Good idea to use matrix here 👍

File: .github/actions/stringify_workflow/action.yml
Comment (XComp) [inline .github/actions/stringify_workflow/action.yml:19]: I feel we can make this more generic and call it just "Stringify". There should be only one input variable `value` and one output `stringified_value`. That makes the action open for future use-cases other than workflow names. WDYT?

**Review by XComp — COMMENTED** [review]
Thanks @morazow. Good idea to use matrix for the OS. I have a few minor comments. PTAL

We should also do a test run on your fork before merging the changes (you could comment parts of the job configuration out that are not required to optimize the runtime for the test).

File: .github/workflows/nightly.yml
Comment (morazow) [reply .github/workflows/nightly.yml:104]: Ahh yes correct, removing

File: .github/actions/stringify_workflow/action.yml
Comment (morazow) [reply .github/actions/stringify_workflow/action.yml:19]: Yes agreed, I will update it

File: .github/workflows/template.flink-ci.yml
Comment (morazow) [inline .github/workflows/template.flink-ci.yml:84]: @XComp, I kept this `outputs.stringified-workflow-name` as is, it is used in the later jobs. 

Initially I thought this could be confusing, but maybe all right to keep it as is, the `stringified_value` will be used only within the compile steps

File: .github/workflows/nightly.yml
Comment (morazow) [reply .github/workflows/nightly.yml]: I think we still need the os parameter, apparently include is only for extending the existing matrix setting

Without os param run: https://github.com/morazow/flink/actions/runs/9206750374

With both os and include: https://github.com/morazow/flink/actions/runs/9207285570

Comment (XComp) [reply .github/workflows/nightly.yml]: How come it works without it in the matrix definition of `template.flink-ci.yml` that only uses `include` in the config? What's the difference there? 🤔

**Comment by morazow** [comment]
Test run of Python wheel build: https://github.com/morazow/flink/actions/runs/9207285570

File: .github/workflows/nightly.yml
Comment (morazow) [reply .github/workflows/nightly.yml:104]: From doc it extends if there is already matrix definition: https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations

But I think I got your suggestion, to make it clear let me define similar include based matrix

**Comment by morazow** [comment]
Test run with include matrix setting: https://github.com/morazow/flink/actions/runs/9207684232

Let's keep it like this it more understandable

File: .github/workflows/nightly.yml
Comment (morazow) [reply .github/workflows/nightly.yml]: Test run with include matrix setting: https://github.com/morazow/flink/actions/runs/9207684232

Let's keep it with only include, latest commit, it's more understandable

File: .github/workflows/nightly.yml
Comment (XComp) [inline .github/workflows/nightly.yml:110]: Looks like the python version also got changed in [FLINK-34582](https://issues.apache.org/jira/browse/FLINK-34582) 🤔 We should check why we don't upgrade the Python version for Linux.

File: .github/workflows/nightly.yml
Comment (XComp) [inline .github/workflows/nightly.yml:126]:
```suggestion
      - name: "Build python wheels for ${{ matrix.os_name }}"
```
I feel like using the `os_name` is less tied to GHA here. WDYT?

File: .github/workflows/wheels.yml
Comment (XComp) [inline .github/workflows/wheels.yml:19]: It's better to comment out certain parts of the nightly workflow that are not needed (test runs and packaging/licensing jobs) and run the actual nightly workflow with the wheels creation. That way we actually test the code rather than some other workflow that would be deleted in the end.

Commenting out of parts of the nightly workflow can happen in a separate commit which can be easily dropped before merging the PR.

**Review by XComp — COMMENTED** [review]
Thanks again, @morazow. I added some comments below. PTAL

File: .github/workflows/wheels.yml
Comment (morazow) [reply .github/workflows/wheels.yml:19]: Ahh got it, on my fork the nightly build was all grayed out. I thought this was due to some secret requirement from original repo. But it is a dispatch workflow, I will try with what suggested after addressing other suggestions

File: .github/workflows/nightly.yml
Comment (morazow) [reply .github/workflows/nightly.yml:99]: We do that in the `tar` command for `flink-python/dist`, right?

Both os version build and create this folder that then we archive and upload

File: .github/workflows/nightly.yml
Comment (morazow) [reply .github/workflows/nightly.yml:110]: Yes, indeed, the linux version of build scripts uses versions until 3.10 but not 3.11. I have asked on the ticket.

https://github.com/HuangXingBo/flink/blob/master/flink-python/dev/build-wheels.sh#L19-L26

File: .github/workflows/nightly.yml
Comment (morazow) [reply .github/workflows/nightly.yml:99]: Ahh yeah, from the ticket there is no folders so we may need to change tar to strip parent folder with `-C` option.

From test runs:

Screenshot 2024-05-23 at 16 35 55

**Comment by morazow** [comment]
Artifacts from new nightly run:

Screenshot 2024-05-23 at 17 21 10

File: .github/workflows/nightly.yml
Comment (XComp) [reply .github/workflows/nightly.yml]: > You're checking [Xingbo Huang](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hxb)'s fork where the master branch doesn't seem to be up-to-date. [apache/flink:flink-python/dev/build-wheels.sh](https://github.com/apache/flink/blob/master/flink-python/dev/build-wheels.sh#L19-L26) does, indeed, have 3.11 added to the python version list.

Ok, I [responded in the Jira issue](https://issues.apache.org/jira/browse/FLINK-34582?focusedCommentId=17849185&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17849185).

Looks like the Python version doesn't make a diffrence here for Linux. We might want to add this as a comment and refer to the build script stating that the version is specified there.

In the mean time: Can you find out where the Python versions are configured for the MacOS builds? It appears that we also have the very same list of Python versions used for building macos wheels (see [this Azure pipeline run|https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=59794&view=logs&j=f73b5736-8355-5390-ec71-4dfdec0ce6c5&t=90f7230e-bf5a-531b-8566-ad48d3e03bbb&l=49] as an example). I only found [apache/flink:flink-python/pyproject.toml:30|https://github.com/apache/flink/blob/942d234444636b13df7a118274d307943f5524cd/flink-python/pyproject.toml#L30]. That might be the right configuration. But I don't understand why the Python version needed to be updated in the workflow yaml to 3.11 in FLINK-34582 in that case. We might need a comment there to clarify things for macos as well.

File: .github/workflows/nightly.yml
Comment (XComp) [reply .github/workflows/nightly.yml:99]: I'm not talking about the subfolder here. I mean the actual build: the wheels build doesn't depend on the compile workflow right now, i.e. it doesn't wait for Flink to be build. Therefore, the Flink build (or rather the relevant Flink dependency) isn't properly packaged with the wheels. ...as far as I understand.

Comment (morazow) [reply .github/workflows/nightly.yml:99]: Hey @XComp,

I think this is fine. The Azure Pipelines (https://github.com/apache/flink/blob/master/tools/azure-pipelines/build-python-wheels.yml) also run like that, builds each wheel in a separate job. 

From [GitHub migration doc](https://docs.github.com/en/actions/migrating-to-github-actions/manually-migrating-to-github-actions/migrating-from-azure-pipelines-to-github-actions#migrating-jobs-and-steps):

```
- Jobs contain a series of steps that run sequentially.
- Jobs run on separate virtual machines or in separate containers.
- Jobs run in parallel by default, but can be configured to run sequentially.
```

So should be fine to build each wheel separately without the Flink build similar to the Azure Pipelines.

File: .github/workflows/nightly.yml
Comment (morazow) [reply .github/workflows/nightly.yml]: Yes on macos the only pyproject.toml is used by cibuildwheel. Tools seems to be well documented, maybe we could also use it for linux builds.

It uses the `cp38-*`, etc for Python versions, https://cibuildwheel.pypa.io/en/stable/options/#build-skip. For Linux we are running bash scripts for building wheel, but we could let the cibuildwheel action do for both platforms.

We can ask @HuangXingBo for final review

**Comment by morazow** [comment]
Hey @XComp, @HuangXingBo ,

What do you think of the latest commit changes for migration?

This way both Linux & MacOS platform use similar build system, we also don't depend on bash script on Linux. All the wheel build requirements are defined in the `pyproject.toml` (python versions, architecture, etc) that are used for both platforms.

For Linux I had to add `manylinux2014` version since `manylinux1` does not support python3.10+ versions.

Please let me know what you think

**Comment by morazow** [comment]
Test run built wheel artifacts: https://github.com/morazow/flink/actions/runs/9224143298

**Review by HuangXingBo — APPROVED** [review]
Thanks the great work of building Python Wheels migration. LGTM.

**Comment by morazow** [comment]
Hey @XComp, could you please have look to the PR?

File: .github/workflows/nightly.yml
Comment (XComp) [inline .github/workflows/nightly.yml:121]: Interesting approach. It's just that Apache doesn't allow actions other than the ones from the `apache`, `github` and `actions` ([source](https://infra.apache.org/github-actions-policy.html)). We could use a fixed version for the custom action and review that specific version. But `pypa/cibuildwheel@v2.16.5` seems to be a bigger project. I'm wondering whether it's worth it or just use the previous approach (even though I liked your intention 👍 ). WDYT?

Comment (morazow) [reply .github/workflows/nightly.yml:121]: Ahh I see. I would be fine to use generic install aproach, right?

```
      # Used to host cibuildwheel
      - uses: actions/setup-python@v5

      - name: Install cibuildwheel
        run: python -m pip install cibuildwheel==2.18.1

      - name: Build wheels
        run: python -m cibuildwheel --output-dir wheelhouse
```

cibuildwheel docs also suggest this generic way for GitHub actions

Comment (XComp) [reply .github/workflows/nightly.yml:121]: That should work 👍

Comment (morazow) [reply .github/workflows/nightly.yml]: Great, updated ✔️

Comment (XComp) [reply .github/workflows/nightly.yml]: Looks fine from my side. But I am not familiar with the whole buildwheel logic. @HuangXingBo can you do another pass over it and approve the changes once more?

**Comment by HuangXingBo** [comment]
I downloaded the wheel package from https://github.com/morazow/flink/actions/runs/9224143298 and found that the wheel package for Linux is manylinux2014, not manylinux1, which may cause it to fail to install on some old machines. I used `dev/glibc_version_fix.sh` and `patchelf` in build-wheels.sh to solve this problem previously.

**Comment by morazow** [comment]
Thanks @HuangXingBo for the update!

I have changed the GitHub actions to run the `auditwheel` repair after installing `patchelf` in Linux. Could you please have another look?

You can find the build wheel artifacts for this change here: https://github.com/morazow/flink/actions/runs/9464510932

**Comment by HuangXingBo** [comment]
Thanks @morazow for the update.
I downloaded the wheel package and found that the wheel package after repair is manylinux_2_5. It may be related to not using `dev/glibc_version_fix.sh`. If the version of glibc is too new, it is impossible to generate manylinux1 package.

File: .github/workflows/nightly.yml
Comment (HuangXingBo) [inline .github/workflows/nightly.yml:31]: Is this code no longer needed? Why not just delete it?

Comment (XComp) [reply .github/workflows/nightly.yml:31]: That's just disabled code to fasten the test run. It will be reverted before merging the PR.

Comment (morazow) [reply .github/workflows/nightly.yml:31]: Yes, I will put it back once we finalize the PR, commented just to run to create the wheels.

**Comment by morazow** [comment]
Thanks @HuangXingBo for the feedback!

Should we still support `manylinux1`? It seems already end of life for some years

**Comment by HuangXingBo** [comment]
@morazow Since a lot of our dependencies are manyLinux2014, I think we can indeed build manyLinux2014.

**Comment by morazow** [comment]
Hello @HuangXingBo, thanks for the feedback!

Is the current state fine? It uses the `manylinux2014` for Linux:

```
[tool.cibuildwheel.linux]
archs = ["x86_64"]
manylinux-x86_64-image = "manylinux2014"
```

Should I also keep the `auditwheel` steps?

**Comment by HuangXingBo** [comment]
@morazow I am not sure whether our wheel package can be manylinux2014 without auditwheel. We can try to remove auditwheel to see the effect

**Comment by morazow** [comment]
Hey @HuangXingBo,

I think one of the runs above was with `manylinux2014` and without auditwheel step. But I will do another workflow run to create the wheels. I'll let you know

**Comment by morazow** [comment]
Hey @HuangXingBo,

I have run two builds, one is with default using `auditwheel` (together with patchelf) and another without `auditwheel`. Both on the `manylinux2014` on Linux.

- With auditwheel: https://github.com/morazow/flink/actions/runs/9566314731
- Without auditwheel: https://github.com/morazow/flink/actions/runs/9566526379

I also updated the build workflow to make explicit some of the steps, previously it was running auditwheel by default on linux.

**Comment by HuangXingBo** [comment]
@morazow  Thanks a lot for the experiment. It seems that we really need auditwheel, otherwise we can't generate manylinux packages.

**Comment by morazow** [comment]
Hey @HuangXingBo, thanks for looking into it 👍 

I have added the `auditwheel repair` step and rebased. Please have a final look if all is fine.

**Review by HuangXingBo — APPROVED** [review]
LGTM

**Comment by XComp** [comment]
Thanks for your help @HuangXingBo and @morazow for your persistence. :-) Code-wise, I'm fine after @HuangXingBo approved the functionality. 👍 

@morazow can we get a final test run of the workflow in your fork (feel free to comment the extraneous parts of the workflow out in the debug commit)? We have to wait for the release branch cutting anyway.

**Comment by morazow** [comment]
Hey all,

Indeed, it took many conversations :)) But it was helpful collaborations! Thank a lot!

@XComp yes, this is the run with only Python wheel build: https://github.com/morazow/flink/actions/runs/9595245039. I will drop the comment commit and rebase once it is okay from your side also 👍

**Review by XComp — APPROVED** [review]

**Comment by morazow** [comment]
Great, thanks @XComp, @HuangXingBo for the help 👍

**Pull request merged**
